### PR TITLE
NXJS-204: handle OAuth client secret

### PR DESCRIPTION
**Pull request by troger**

### CODE REVIEW

**Review by kevinleturc — APPROVED** [review]

**Review by kevinleturc — DISMISSED** [review]

File: README.md [hunk 1]
```diff
@@ -305,45 +345,64 @@ The Nuxeo Platform replies with an access token:
 
 There are some utility methods on the client to help you with this flow:
 
-__Nuxeo.oauth2.getAuthorizationURL(baseURL, clientId[, params])__
+__Nuxeo.oauth2.getAuthorizationURL(opts)__
 
-Returns the OAuth2 authorization URL for the given `baseURL` and `clientId`.
+Returns the OAuth2 authorization URL for the given `opts` object, containing `baseURL`, `clientId` and optional `clientSecret` and `params`.
```
Comment (kevinleturc) [inline README.md:350]:
```suggestion
Returns the OAuth2 authorization URL for the given `opts` object, containing `baseURL`, `clientId`, and optional `clientSecret` and `params`.
```

File: README.md [hunk 2]
```diff
@@ -305,45 +345,64 @@ The Nuxeo Platform replies with an access token:
...
+  }
 });
 console.log(authorizationURL); // http://localhost:8080/nuxeo/oauth2/authorize?client_id=my-app&response_type=code&state=xyz&redirect_uri=http://localhost:8000/authorize
 ```
 
-__Nuxeo.oauth2.fetchAccessTokenFromAuthorizationCode(baseURL, clientId, code[, params])__
+__Nuxeo.oauth2.fetchAccessTokenFromAuthorizationCode(opts)__
 
-Fetches an OAuth2 access token for the given `baseURL`, `clientId` and `code`.
+Fetches an OAuth2 access token for the given `opts` object containing `baseURL`, `clientId`, `code` and optional `clientSecret` and `params`.
```
Comment (kevinleturc) [inline README.md:367]:
```suggestion
Fetches an OAuth2 access token for the given `opts` object containing `baseURL`, `clientId`, `code`, and optional `clientSecret` and `params`.
```

File: README.md [hunk 3]
```diff
@@ -305,45 +345,64 @@ The Nuxeo Platform replies with an access token:
...
     }
   });
 });
 ```
 
-__Nuxeo.oauth2.fetchAccessTokenFromJWTToken(baseURL, clientId, jwtToken[, params])__
+__Nuxeo.oauth2.fetchAccessTokenFromJWTToken(opts)__
 
-Fetches an OAuth2 access token for the given `baseURL`, `clientId` and `jwtToken`.
+Fetches an OAuth2 access token for the given `opts` object containing `baseURL`, `clientId`, `jwtToken` and optional `clientSecret` and `params`.
```
Comment (kevinleturc) [inline README.md:394]:
```suggestion
Fetches an OAuth2 access token for the given `opts` object containing `baseURL`, `clientId`, `jwtToken`, and optional `clientSecret` and `params`.
```

File: README.md [hunk 4]
```diff
@@ -355,14 +414,18 @@ Nuxeo.oauth2.fetchAccessTokenFromJWTToken('http://localhost:8080/nuxeo', 'my-app
 });
 ```
 
-__Nuxeo.oauth2.refreshAccessToken(baseURL, clientId, refreshToken[, params])__
+__Nuxeo.oauth2.refreshAccessToken(optsbaseURL, clientId, refreshToken[, params])__
 
-Manually refresh an access token for the given `baseURL`, `clientId` and `refreshToken`.
+Manually refresh an access token for the given `opts` object containing `baseURL`, `clientId`, `refreshToken` and optional `clientSecret` and `params`.
```
Comment (kevinleturc) [inline README.md:419]:
```suggestion
Manually refresh an access token for the given `opts` object containing `baseURL`, `clientId`, `refreshToken`, and optional `clientSecret` and `params`.
```

File: lib/auth/oauth2.js [hunk 5]
```diff
@@ -69,6 +90,7 @@ const oauth2 = {
    * @param {string} [code] - An authorization code. Deprecated since 3.16.0, use other helper methods instead.
    * @param {object} [params] - Optional parameters such as `redirect_uri`.
    * @returns {string}
+   * @deprecated since 3.16.0, use other helper methods instead.
```
Comment (kevinleturc) [inline lib/auth/oauth2.js:93]:
```suggestion
   * @deprecated since 4.0.0, use other helper methods instead.
```
?

**Review by kevinleturc — APPROVED** [review]

File: lib/auth/oauth2.js (hunk 5 above)
Comment (troger) [reply lib/auth/oauth2.js:93]: The code arg was deprecated on 3.16.0 (see jsdoc), so let's deprecate the whole method also from the same version.

File: README.md (hunk 1 above)
Comment (troger) [reply README.md:350]: I'm not sure, and anyway I'm not really fan of the 2 `and`s... Waiting for @ataillefer review to have a better wording 👍

File: README.md (hunk 2 above)
Comment (ataillefer) [reply README.md]: containing `baseURL`, `clientId`, `code`, optional `clientSecret`, (and)  `params`

File: README.md (hunk 3 above)
Comment (ataillefer) [reply README.md]: containing `baseURL`, `clientId`, `jwtToken` optional `clientSecret` (and) `params`

File: README.md (hunk 4 above)
Comment (ataillefer) [reply README.md]: baseURL`, `clientId`, `refreshToken`, optional `clientSecret`, (and) `params`

File: lib/auth/auth.js [hunk 6]
```diff
@@ -125,7 +125,8 @@ const bearerTokenAuthenticator = {
         return resolve(auth);
       }
 
-      return oauth2.refreshAccessToken(baseURL, auth.clientId, auth.token.refresh_token)
+      const params = typeof auth.clientSecret !== 'undefined' ? { client_secret: auth.clientSecret } : {};
+      return oauth2.refreshAccessToken(baseURL, auth.clientId, auth.token.refresh_token, params)
```
Comment (ataillefer) [inline lib/auth/auth.js:129]: Is it the deprecated method?

File: lib/auth/oauth2.js [hunk 7]
```diff
@@ -39,27 +39,48 @@ const extractCodeParams = (args) => {
...
 
   /**
    * Returns the OAuth2 authorization URL.
-   * @param {string} [baseURL] - Base URL of the Nuxeo Platform.
-   * @param {string} [clientId] - The OAuth2 client id.
-   * @param {object} [opts] - Optional query parameters such as `state`, `redirect_uri`.
+   * @param {string} opts - The configuration options.
+   * @param {string} opts.baseURL - Base URL of the Nuxeo Platform.
+   * @param {string} opts.clientId - The OAuth2 client id.
+   * @param {string} [opts.clientSecret] - Optional OAuth client secret
```
Comment (ataillefer) [inline lib/auth/oauth2.js:66]:
```suggestion
   * @param {string} [opts.clientSecret] - Optional OAuth2 client secret.
```

File: lib/auth/oauth2.js [hunk 8]
```diff
@@ -88,64 +110,91 @@ const oauth2 = {
...
   /**
    * Fetches an OAuth2 access token from an authorization code.
-   * @param {string} baseURL - Base URL of the Nuxeo Platform.
-   * @param {string} clientId - The OAuth2 client id.
-   * @param {string} code - An authorization code.
-   * @param {object} [params] - Optional parameters such as `redirect_uri`.
+   * @param {string} opts - The configuration options.
+   * @param {string} opts.baseURL - Base URL of the Nuxeo Platform.
+   * @param {string} opts.clientId - The OAuth2 client id.
+   * @param {string} [opts.clientSecret] - Optional OAuth client secret
```
Comment (ataillefer) [inline lib/auth/oauth2.js:116]:
```suggestion
   * @param {string} [opts.clientSecret] - Optional OAuth2 client secret.
```

File: lib/auth/oauth2.js [hunk 9]
```diff
@@ -88,64 +110,91 @@ const oauth2 = {
...
   /**
    * Fetches an OAuth2 access token from a JWT token.
-   * @param {string} baseURL - Base URL of the Nuxeo Platform.
-   * @param {string} clientId - The OAuth2 client id.
-   * @param {string} jwtToken - A JWT token.
-   * @param {object} [params] - Optional parameters such as `redirect_uri`.
+   * @param {string} opts - The configuration options.
+   * @param {string} opts.baseURL - Base URL of the Nuxeo Platform.
+   * @param {string} opts.clientId - The OAuth2 client id.
+   * @param {string} [opts.clientSecret] - Optional OAuth client secret
```
Comment (ataillefer) [inline lib/auth/oauth2.js:145]:
```suggestion
   * @param {string} [opts.clientSecret] - Optional OAuth2 client secret.
```

File: lib/auth/oauth2.js [hunk 10]
```diff
@@ -88,64 +110,91 @@ const oauth2 = {
...
   /**
    * Refreshes an OAuth2 access token.
-   * @param {string} [baseURL] - Base URL of the Nuxeo Platform.
-   * @param {string} [clientId] - The OAuth2 client id.
-   * @param {string} [refreshToken] - A refresh token.
-   * @param {object} [params] - Optional parameters.
+   * @param {string} opts - The configuration options.
+   * @param {string} opts.baseURL - Base URL of the Nuxeo Platform.
+   * @param {string} opts.clientId - The OAuth2 client id.
+   * @param {string} [opts.clientSecret] - Optional OAuth client secret
```
Comment (ataillefer) [inline lib/auth/oauth2.js:175]:
```suggestion
   * @param {string} [opts.clientSecret] - Optional OAuth2 client secret.
```

File: README.md (hunk 1 above)
Comment (troger) [reply README.md]: Went for:
```
Returns the OAuth2 authorization URL for the given `opts` object, containing `baseURL`, `clientId`, optional `clientSecret` and `params`.
```

**Review by ataillefer — APPROVED** [review]

File: lib/auth/auth.js (hunk 6 above)
Comment (troger) [reply lib/auth/auth.js:129]: Good catch!

**Review by ataillefer — APPROVED** [review]

**Review by kevinleturc — APPROVED** [review]

**Pull request merged**